scripts/tagging_scripts/insert_id_and_ref_tags.py

```diff
@@ -0,0 +1,256 @@
+"""
+Utility script to read the list of topics from mongodb and insert mandatory
+id and reference tags for these topics. The script only works for topics
+with the naming convention /campus/building/device/<optional sub device>/point.
+A error will be displayed for any topics that do not conform to this naming
+convention and will be skipped.
+
+**Configure user configuration variables right below imports before running the
+scripts.**
+
+By default the script only set id field and parent references but if there are
+other common tags that apply to ALL campus, building, device, or point, you
+could update the tags variable in get_campus_tags, get_site_tags,
+get_equip_tags, and get_point_tags method respectively
+"""
+
+import random
+import re
+import sqlite3
+from pymongo import MongoClient
+from pymongo.errors import BulkWriteError
+import datetime
+
+## User configuration - start ##
+
+#Connection string to the mongo db from which list of topics should be read
+topics_connection_string = "mongodb://reader:volttronReader@vc-db.pnl.gov" \
+                           ":27017" \
+                           "/prod_historian"
+topic_table = "topics"
+# Regular expression that matches all device topics that conform to the
+# naming pattern /campus/building/device/<optional sub device>/point
+device_topics = "^PNNL"
+
+# table or collection name into which tags should be saved.
+tags_table = "topic_tags_4"
+
+# The connection details of the database into which tag tables are to be
+# created.
+# set mongo, sqlite or both. If you need tags to be saved in mongodb you can
+#  set sqlite connection string as empty/None and vice-versa
+tags_db_mongo_conn_str = "mongodb://test:test@localhost:27017/mongo_test"
+tags_db_sqlite = "/home/velo/tags_test3.sqlite"
+
+## User configuration - end ##
+
+write_to_mongo = False
+if tags_db_mongo_conn_str:
+    topics_mongodb = MongoClient(topics_connection_string).get_default_database()
+    tags_client = MongoClient(tags_db_mongo_conn_str)
+    tags_mongodb = tags_client.get_default_database()
+    mongo_bulk = tags_mongodb[tags_table].initialize_ordered_bulk_op()
+    mongo_batch_size = 0
+    mongo_max_batch_size = 5000
+    write_to_mongo = True
+
+write_to_sqlite = False
+if tags_db_sqlite:
+    sqlite_connection = sqlite3.connect(
+        tags_db_sqlite,
+        detect_types = sqlite3.PARSE_DECLTYPES | sqlite3.PARSE_COLNAMES
+    )
+    sqlite_connection.execute("PRAGMA CACHE_SIZE=-2000")
+
```

Comment:
Why a - cache size?

Reply:
without a large enough cache sqlite queries can be super slow.

Reply:
Yes but why is it negative!

Reply:
it indicates it is in kilobytes and not pages.

Reply:
Ahh...that would be a good spot for a comment...as I had no clue.